### PR TITLE
Unify `end2end` under `nms=True|False|None`

### PR DESCRIPTION
**Pull request by glenn-jocher**
## Summary

Consolidate inference head selection and export NMS under the existing `nms` argument, deprecating `end2end`. YOLO models with both heads now default to one-to-many predictions with NMS using the same checkpoint. Bump the package version to `8.4.142`.

| `nms` | Predict, val and track | Export |
| --- | --- | --- |
| `None` (default) | One-to-many with external NMS | Raw one-to-many outputs |
| `True` | One-to-many with NMS | Embed NMS where supported |
| `False` | NMS-free one-to-one head when available | NMS-free outputs where supported |

Both heads retain their training supervision. Epoch validation, checkpoint selection and early stopping follow the selected inference head. Models without selectable heads and existing exported graphs retain their native output behavior.

Legacy `end2end=True` maps to `nms=False`; `end2end=False` maps to `nms=None` unless paired with `nms=True`. Existing conflicting pairs retain their previous head-selection precedence. Explicit `nms=True` now selects the one-to-many head instead of being ignored on dual-head models.

Head selection occurs before native fusion for both checkpoint filenames and loaded modules. Shared head fusion removes the unused branch and replaces duplicate subclass implementations. Predictor reuse and benchmarks use the same selection. Documentation presents nms as the canonical interface without migration notes; backward compatibility is handled by the configuration deprecation handler. Argument tables and existing export matrices cover the unified contract. TensorRT execution reads current tensor pointers after dynamic output resizing, fixing a stale-buffer error exposed by the larger one-to-many outputs on TensorRT 8.

## Validation

- 240 existing slow ONNX/TorchScript matrix cases passed locally across tasks, NMS modes, dynamic shapes and batches.
- Focused training, validation, resume, distillation and native inference checks passed, including both training losses and one-to-many epoch validation.
- Existing YOLOE prompt-free checkpoints, vocabulary conversion, legacy arguments and reused predictors checked through real model paths.
- CoreML embedded-NMS export and native reload/inference passed with the repository export extra.
- The known-assets registry adds 19 names used by YOLOE, tracking, or documented checkpoint workflows in five lines: MobileCLIP2, ReID ONNX, distilled and ADE20K checkpoints, YOLOv9 segmentation, and YOLOv8 pose-p6. Each name was checked against the published release. Fresh MobileCLIP2 download/text encoding, distilled-checkpoint inference, and ReID auto-download/inference through the tracking owner passed.
- The COCO8-pose benchmark floor follows the selected head: unmodified main gives 0.1795 for one-to-many versus 0.18516 for one-to-one at imgsz=160. The floor is now 0.179 so CI exercises the new default; other floors are unchanged.
- Hailo host decoding matches real ONNX outputs from both detection heads at the compiler-selected output nodes. Full HEF compilation/device execution was not available locally.
- Strict documentation build and Python 3.8 syntax checks passed.

Claude Code Fable 5.1 high and an independent cold GPT-6 Astra xhigh full-diff review return LGTM with no findings on `edb6fc66f`, including the version-only delta. Automated review threads are resolved.

[Normal CI](https://github.com/ultralytics/ultralytics/actions/runs/33993980663) passed on `edb6fc66f`; the preceding [normal CI run](https://github.com/ultralytics/ultralytics/actions/runs/33990884798) passed before the version-only change. [Full slow CI](https://github.com/ultralytics/ultralytics/actions/runs/33988342498) and [Docker CI](https://github.com/ultralytics/ultralytics/actions/runs/33988343769) passed on `c16efcee6`; the only subsequent changes are a one-line correction to the distillation benchmark reproduction command and the package version bump. Inference code, tests, assets, and workflows are identical.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

**WARNING ⚠️** this PR is very large, summary may not cover all changes.

### 🌟 Summary

Inference-head selection and export NMS behavior now use the unified `nms=None|True|False` interface, with `end2end` deprecated and YOLO26 defaulting to one-to-many predictions with NMS.

### 📊 Key Changes

- `nms=None` selects one-to-many outputs with external NMS, `nms=True` embeds NMS where supported, and `nms=False` selects the NMS-free one-to-one head when available across prediction, validation, tracking, benchmarking, and export.
- Training retains both one-to-many and one-to-one heads and losses; epoch validation, checkpoint selection, and early stopping use the head selected by `nms`.
- Legacy `end2end` arguments are handled by the deprecation layer: `True` maps to `nms=False`, while `False` maps to `nms=None` unless `nms=True` is explicitly set.
- Head selection occurs before fusion, and fusion removes the unused branch; predictor reuse, native inference, Hailo output selection, and export paths use the same selection logic.
- Documentation, export argument tables, tests, and CI benchmarks were updated for the unified interface, and the package version was bumped to `8.4.142`.

### 🎯 Purpose & Impact

- YOLO26 prediction and validation now use the accuracy-oriented one-to-many head with NMS by default; users can request NMS-free inference with `nms=False` or embedded NMS during supported exports with `nms=True`.
- Existing workflows using `end2end` continue through deprecation mapping, while `end2end` is removed from the documented configuration and export API.
- Exported output behavior is now explicit: raw one-to-many outputs by default, embedded NMS where supported, or one-to-one NMS-free outputs with `nms=False`; unsupported formats retain their native output path.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: ultralytics/cfg/__init__.py [hunk 1]
```diff
@@ -369,7 +368,7 @@ def get_cfg(
           `project` and `name` to strings and validating configuration keys and values.
         - The function performs type and value checks on the configuration data.
     """
-    cfg = cfg2dict(cfg)
+    cfg = _handle_deprecation(cfg2dict(cfg))
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/__init__.py:371]: 💡 **MEDIUM**: `_handle_deprecation` is applied only to the base `cfg`, not to `overrides`. Therefore calls such as `get_cfg(overrides={"end2end": False})` still fail `check_dict_alignment` because `end2end` was removed from the default configuration. This breaks the promised legacy compatibility for `val`, `export`, and direct predictor/validator construction; normalize the overrides before alignment as well.

File: ultralytics/cfg/__init__.py [hunk 2]
```diff
@@ -1053,10 +1060,10 @@ def entrypoint(debug: str = "") -> None:
         elif a.lower() in special:
             special[a.lower()]()
             return
-        elif a in DEFAULT_CFG_DICT and isinstance(DEFAULT_CFG_DICT[a], bool):
-            overrides[a] = True  # auto-True for default bool args, i.e. 'yolo show' sets show=True
-        elif a in {"half", "int8"}:
-            overrides[a] = True  # deprecated bare precision flags, forwarded to quantize by _handle_deprecation
+        elif a in CFG_BOOL_KEYS:
+            overrides[a] = True  # auto-True for bool args, i.e. 'yolo show' sets show=True
+        elif a in {"half", "int8", "end2end"}:
```
Comment (UltralyticsAssistant) [inline ultralytics/cfg/__init__.py]: 💡 **MEDIUM**: This accepts a bare `end2end` token but stores the deprecated key unchanged in `overrides`; the subsequent `check_dict_alignment(full_args_dict, overrides)` rejects it because `end2end` is no longer in `full_args_dict`. Explicit CLI forms such as `end2end=True` fail even earlier for the same reason. Normalize the deprecated CLI argument before the final key check, or include it in the accepted legacy-key handling.

File: ultralytics/nn/tasks.py [hunk 3]
```diff
@@ -1306,8 +1311,8 @@ def set_vocab(self, vocab, names):
         with torch.no_grad():  # a tracked warmup would build a graph through the backbone
             self(next(self.parameters()).new_empty(1, 3, self.args["imgsz"], self.args["imgsz"]))  # warmup
 
-        cv3 = getattr(head, "one2one_cv3", head.cv3)
-        cv2 = getattr(head, "one2one_cv2", head.cv2)
+        cv3 = head.one2one_cv3 if head.end2end else head.cv3
```
Comment (UltralyticsAssistant) [inline ultralytics/nn/tasks.py:1314]: 💡 **MEDIUM**: `set_vocab()` reparameterizes only the currently selected branch and removes that branch's final layers. After this call, switching `nms` can select the untouched other branch, while `forward_lrpc()` still applies the LRPC modules built for the reparameterized branch; its proposal/localization convolutions then receive post-final or otherwise incompatible feature channels. This affects prompt-free YOLOE usage such as `set_vocab(...); predict(nms=False)`. Reparameterize both selectable branches or keep branch-specific LRPC modules and switch them with the selected head.

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

The PR has concrete compatibility and YOLOE prompt-free inference regressions. Legacy `end2end` arguments are not normalized in general configuration or CLI overrides, and `set_vocab()` is not safe when the newly selectable YOLOE heads are switched after reparameterization. These should be fixed before merge.

<details><summary>💬 Posted 3 inline comments</summary>

- 💡 **MEDIUM** `ultralytics/cfg/__init__.py:371` `_handle_deprecation` is applied only to the base `cfg`, not to `overrides`. Therefore calls such as `get_cfg(overrides={"end2end": False})` still fail `check_dict_alignment` because `end2end` was removed from the default configuration. This breaks the promised legacy compatibility for `val`, `export`, and direct predictor/validator construction; normalize the overrides before alignment as well.
- 💡 **MEDIUM** `ultralytics/cfg/__init__.py:1065` This accepts a bare `end2end` token but stores the deprecated key unchanged in `overrides`; the subsequent `check_dict_alignment(full_args_dict, overrides)` rejects it because `end2end` is no longer in `full_args_dict`. Explicit CLI forms such as `end2end=True` fail even earlier for the same reason. Normalize the deprecated CLI argument before the final key check, or include it in the accepted legacy-key handling.
- 💡 **MEDIUM** `ultralytics/nn/tasks.py:1314` `set_vocab()` reparameterizes only the currently selected branch and removes that branch's final layers. After this call, switching `nms` can select the untouched other branch, while `forward_lrpc()` still applies the LRPC modules built for the reparameterized branch; its proposal/localization convolutions then receive post-final or otherwise incompatible feature channels. This affects prompt-free YOLOE usage such as `set_vocab(...); predict(nms=False)`. Reparameterize both selectable branches…
</details>

File: ultralytics/cfg/__init__.py (hunk 1 above)
Comment (glenn-jocher) [reply ultralytics/cfg/__init__.py:371]: Verified through get_cfg(overrides={"end2end": False}) and the existing configuration regression test. check_dict_alignment() already calls _handle_deprecation(custom) before comparing keys, so this path accepts and translates the legacy override. Adding another translation call would duplicate the existing owner.

File: ultralytics/cfg/__init__.py (hunk 2 above)
Comment (glenn-jocher) [reply ultralytics/cfg/__init__.py]: Verified the exact CLI forms end2end=True, end2end=False and bare end2end through real prediction. The final check_dict_alignment() translates deprecated keys before comparing them. The follow-up commit also preserves all existing bare boolean defaults while accepting bare nms.

File: ultralytics/nn/tasks.py (hunk 3 above)
Comment (glenn-jocher) [reply ultralytics/nn/tasks.py:1314]: Fixed in ab7a2b047. Vocabulary conversion now calls the existing head.fuse() to remove the unused branch: LRPC has already irreversibly consumed the selected branch. The existing head-availability fallback then keeps inference on that branch when a conflicting selection is requested. A real YOLOE vocabulary-conversion regression test reproduces the previous channel mismatch and passes with the fix, without adding branch-specific LRPC copies.

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-09-05 16:56:20 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/33979344329): `ci.yml` run 31206
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/33979345282): `docker.yml` run 3171

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-09-05 17:00:26 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/33979540898): `ci.yml` run 31208
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/33979541544): `docker.yml` run 3172

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2026-09-05 19:52:00 UTC with `@ultralytics/run-all` command
(available commands are `@ultralytics/run-all`, `@ultralytics/run-ci`, and `@ultralytics/run-docker`):

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/33988342498): `ci.yml` run 31229
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/33988343769): `docker.yml` run 3174

**Comment by UltralyticsAssistant** [comment]
🎉 Thank you, @glenn-jocher, for merging [PR #26066](https://github.com/ultralytics/ultralytics/pull/26066)!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This PR brings together a unified `nms=None|True|False` interface across prediction, validation, tracking, benchmarking, and export, making YOLO26 head selection and output behavior more explicit while preserving legacy `end2end` workflows through deprecation mapping.
